### PR TITLE
Pie 2675/generate notifications

### LINKED ISSUE #2675: Generate notifications when a child's approval is about to expire
## 📋 Summary

When any approval is within 30 days of expiration, we should generate a notification that can be accessed via the API.
When a child has a future approval that is more than 30 days expiration

## 🇺🇸 States this change applies to
- [X] ALL
- [ ] Nebraska
- [ ] Illinois

## 💻 Implementation Details

- ~~a second~~ option that's a little more fully featured:
  - add a `notifications` table where each record `belongs_to` a `child` and an `approval`
  - run a job every day to create notifications for approvals that are expiring within 30 days, where no future approval exists
  - run a job every day (maybe the same job) to clear out notifications that have been "resolved"
  - add a `notifications` endpoint (with the same policies as other controller actions)
  - add a request to the `notifications` endpoint in the notifications component already constructed
- pros: fuller functionality, can clear out notifications in the job when a child's approval is renewed in the future, performance hit of front-end parsing is eliminated
- cons: adds an additional API hit on the Dashboard, might require a loading graphic if performance is slow, this is a section of the data model we haven't incorporated in depth and might change later

per discussion with @rebeccakarasiktw and @katelovescode we want to add performance testing with large #s of children as a part of implementing this ticket

Archived notes/discussion:

> There are a couple ways we can manage this:
> - ~~the `case_list_for_dashboard` endpoint is *already* returning the information needed to generate these messages; we return `first_name` and `last_name` in the `ChildBlueprint`, and `approval_expires_on` in the nested `NebraskaDashboardCaseBlueprint`.  The front-end *could* filter the cases on the dashboard for ones expiring in the next 30 days and generate notifications at load time.~~
>   - ~~pros: no need to change the backend~~
>   - ~~cons: adds more responsibility for parsing logic on the front-end, no clear way to "clear" the messages once the child has been updated, until they switch to the new approval, possible performance hit for front-end filtering of all the cases~~

~~cc @rebeccakarasiktw and @cjhaddad - please have a look at these implementation details and let me know which of these seems like a better option (and which of the functionality I listed above is required vs. which is optional - i.e. clearing out notifications after a new approval is found)~~

~~^^ We want to go with the second option bc it allows for all criteria to be met.~~

### PR DESCRIPTION
**Pull request by OliverCai0**
## 💅🏼 What issue does this fix?
Closes #2675. Adds notification endpoint, daily job to create and update notifications. 

## 🛷 Deployment
* [ ] Migrations
Added a migration to create notification table 20220607193259_create_notifications.rb

## 🧵 Steps to set up locally
rake daily_notifications to queue generate_notification job.

## 🧳 Steps to test
bundle exec rspec spec/services/notification_generator_spec.rb to test notification_generator behavior
bundle exec rspec spec/requests/api/v1/notifications_spec.rb  to test notifications endpoint and scope

## 🕯 Caveats, concerns
In order to achieve the desired behavior of removing notifications once an approval is updated for a child (a new approval record is created), I delete all existing notifications before querying for approvals with the latest expires_on date for each child. This seemed like the most optimal method to me, since I couldn't find a more efficient way to link a child back its latest approval other than an extensive search of all approvals, and I needed to do the query anyways to find the latest approvals.

### CODE REVIEW

File: app/blueprints/approval_blueprint.rb
Comment (katelovescode) [inline app/blueprints/approval_blueprint.rb:3]: Let's delete this blueprint/view since we aren't using it anymore

File: app/blueprints/child_blueprint.rb
Comment (katelovescode) [inline app/blueprints/child_blueprint.rb:59]: Ditto - let's delete this view since we aren't using it anymore

File: app/blueprints/notification_blueprint.rb
Comment (katelovescode) [inline app/blueprints/notification_blueprint.rb:5]: Can we remove this commented-out code?

File: app/services/notification_generator.rb
Comment (katelovescode) [inline app/services/notification_generator.rb:3]: Note: we have the answer to this now - we want one notification per child, let's remove this comment.

File: app/services/notification_generator.rb
Comment (katelovescode) [inline app/services/notification_generator.rb:6]: How might we do this without needing to destroy all the notification records?  What errors could this cause?

As a thought experiment:
- what happens if the job fails in the middle and the Notifications have all been destroyed but the new ones haven't finished generating?

I might suggest a different implementation, like:

- create a variable of all approvals w/ a notification i.e. `approvals_with_notification = Approval.joins(:notification)`
- loop through each of those approvals
  - if the approval is still expiring within 30 days, `return` (do nothing)
  - if not, loop through all the notifications
    - if the notification's child has no approvals after the current approval, `return` (do nothing)
    - if they do have an approval afterwards, delete the notification

- create a variable of approvals expiring within 30 days w/ no notification i.e. `approvals_without_notification = Approval.where.missing(:notification).where('expires_on < ?', 30.days.after`
- if that variable length is greater than 0, loop through each and create a new notification.

Does this make sense?  Let me know what you think.

File: spec/models/notification_spec.rb
Comment (katelovescode) [inline spec/models/notification_spec.rb:6]: Let's write some specs in here!  Even though `Notification` is a tiny model, let's ensure it's properly associated.  Take a look at some of the other `spec/models` files - there will be an example of something like "`it should belong_to X`"

File: spec/support/api/schemas/notification.json
Comment (katelovescode) [inline spec/support/api/schemas/notification.json:2]: The spacing on this JSON is very large, can you make sure you're only using 2 spaces instead of tabs/4+ spaces?

File: spec/support/api/schemas/notifications.json
Comment (katelovescode) [inline spec/support/api/schemas/notifications.json:1]: Ditto spacing comment here

File: app/models/notification.rb
Comment (katelovescode) [inline app/models/notification.rb:6]: Great uniqueness validation here, I hadn't thought about this

File: config/routes.rb
Comment (katelovescode) [inline config/routes.rb:34]: Glad this is scoped only to exactly what we need!

File: db/seeds.rb
Comment (katelovescode) [inline db/seeds.rb:182]: Love this addition to the seeds, thank you!

**Review by katelovescode — CHANGES_REQUESTED** [review]
Really great work, this is a *huge* feature to work on right out of the gate!  I added a couple of suggestions on implementation, and a few requests for some tweaks.  Happy to talk about this on Monday!

File: app/services/notification_generator.rb
Comment (katelovescode) [inline app/services/notification_generator.rb:7]: Handy syntax here:
```
Approval.where.missing(:notifications).where(expires_on: 0.days.after..30.days.after)
```
the range notation `..` works for dates

File: app/models/child.rb
Comment (katelovescode) [inline app/models/child.rb:19]: 😍

File: app/models/child_approval.rb
Comment (katelovescode) [inline app/models/child_approval.rb:12]: Out of curiosity, why are you also associating to a child_approval?  Was this a mistake?  I ask because I don't see you using this association anywhere.

File: app/models/notification.rb
Comment (katelovescode) [inline app/models/notification.rb:6]: How come you decided against this scoped uniqueness?

File: app/services/notification_generator.rb
Comment (katelovescode) [inline app/services/notification_generator.rb:22]: Rather than looping through all Approvals, does it make sense to reduce the number of Approvals you're checking with a join?  `Approval.joins(:notifications).each`?

Comment (katelovescode) [inline app/services/notification_generator.rb:22]: Rather than `record` can we name the param `approval` so it's clear what we're operating on?

File: app/services/notification_generator.rb
Comment (katelovescode) [inline app/services/notification_generator.rb:11]: Can we name this param `approval` rather than `record` for clarity?

File: app/models/notification.rb
Comment (OliverCai0) [reply app/models/notification.rb:6]: I'm not sure if it is still necessary given that there should be only one notification per child since there is only one approval per child.

File: app/services/notification_generator.rb
Comment (OliverCai0) [reply app/services/notification_generator.rb:22]: I wasn't sure if the updated approvals shared the same ID. The approvals that I utilized in the test didn't share id and this join was problematic in the fact that the preexisting notification always linked back to the non updated approval, and I needed to check all the most recent approvals.

File: app/models/child_approval.rb
Comment (OliverCai0) [reply app/models/child_approval.rb]: Yes this was a mistake.

**Pull request merged**
